File: 07.py
import fileinput
import re

def can_contain(rules, color, target):
    return any(c == target or can_contain(rules, c, target)
        for c in rules[color])

def bag_count(rules, color):
    return sum(n + n * bag_count(rules, c)
        for c, n in rules[color].items())

rules = {}
for line in fileinput.input():
    line = line[:-2].replace(' bags', '').replace(' bag', '')
    color, rest = line.split(' contain ')
    rules[color] = {c: int(n)
        for n, c in re.findall(r'(\d+) ([^,]+)', rest)}

t = 'shiny gold'
print(sum(can_contain(rules, c, t) for c in rules))
print(bag_count(rules, t))

File: 07.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type Rule map[string]int
type Rules map[string]Rule

func parseRules(r io.Reader) Rules {
	rules := make(Rules)
	re := regexp.MustCompile(`(\d+) ([^,]+)`)
	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		line := scanner.Text()
		line = strings.TrimSuffix(line, ".")
		line = strings.ReplaceAll(line, " bags", "")
		line = strings.ReplaceAll(line, " bag", "")
		split := strings.Split(line, " contain ")
		rule := make(Rule)
		for _, m := range re.FindAllStringSubmatch(split[1], -1) {
			x, _ := strconv.ParseInt(m[1], 0, 0)
			rule[m[2]] = int(x)
		}
		rules[split[0]] = rule
	}
	return rules
}

func canContain(rules Rules, color, target string) bool {
	for c := range rules[color] {
		if c == target || canContain(rules, c, target) {
			return true
		}
	}
	return false
}

func canContainSum(rules Rules, color string) int {
	var sum int
	for c := range rules {
		if canContain(rules, c, color) {
			sum++
		}
	}
	return sum
}

func bagCount(rules Rules, color string) int {
	var sum int
	for c, n := range rules[color] {
		sum += n + n*bagCount(rules, c)
	}
	return sum
}

func main() {
	rules := parseRules(os.Stdin)
	fmt.Println(canContainSum(rules, "shiny gold"))
	fmt.Println(bagCount(rules, "shiny gold"))
}
